bin/coqtop.ml

```diff
@@ -0,0 +1,81 @@
+      {|$(b,dune coq top FILE -- ARGS) runs the Coq toplevel to process the
+          given $(b,FILE). The given arguments are completed according to the
+          local configuration. This is equivalent to running $(b,coqtop ARGS)
+          with a $(b,_CoqProject) file containing the local configurations
+          from the $(b,dune) files, but does not require maintaining a
+          $(b,_CoqProject) file.|}
+  ; `Blocks Common.help_secs
+  ]
+
+let info = Term.info "top" ~doc ~man
+
+let term =
+  let+ common = Common.term
+  and+ context =
+    Common.context_arg ~doc:{|Run the Coq toplevel in this build context.|}
+  and+ coq_file =
+    Arg.(required & pos 0 (some string) None (Arg.info [] ~docv:"COQFILE"))
+  and+ extra_args =
+    Arg.(value & pos_right 0 string [] (Arg.info [] ~docv:"ARGS"))
+  in
+  let config = Common.init common in
+  (* Checks on the argument file. *)
+  let source = Common.prefix_target common coq_file in
+  if not (Path.is_file (Path.of_string source)) then
+    User_error.raise [ Pp.textf "cannot find file: %s" coq_file ];
+  let coqtop, argv, env =
+    Scheduler.go ~common ~config (fun () ->
+        let open Fiber.O in
+        let* setup = Import.Main.setup () in
+        let* setup = Memo.Build.run setup in
+        let sctx = Import.Main.find_scontext_exn setup ~name:context in
+        let context = Dune_rules.Super_context.context sctx in
+        let dir =
+          let dir = Filename.dirname coq_file in
+          Path.Build.relative context.build_dir (Common.prefix_target common dir)
+        in
+        let* (coqtop, args) =
+          let open Memo.Build.O in
+          Build_system.run_exn (fun () ->
+              let* dc = Dune_rules.Dir_contents.get sctx ~dir in
+              let* coq_src = Dune_rules.Dir_contents.coq dc in
+              let coq_module =
+                let source = Path.Build.relative context.build_dir coq_file in
+                match Dune_rules.Coq_sources.find_module ~source coq_src with
+                | Some m -> m
+                | None -> User_error.raise [ Pp.textf "cannot find file: %s" coq_file ]
+              in
+              let* _ : unit * unit Dep.Map.t =
+                let deps =
+                  Dune_rules.Coq_rules.deps_of ~dir ~boot_type:No_boot coq_module
+                in
+                Action_builder.run deps Lazy
+              in
+              let args =
+                [] (* FIXME arguments used when invoking coqc on the file. *)
+              in
+              let prog = "coqtop" in
+              Super_context.resolve_program sctx ~dir ~loc:None prog
+              >>= function
+              | Error (_ : Action.Prog.Not_found.t) ->
+                  User_error.raise [ Pp.textf "Program %S not found!" prog ]
+              | Ok prog ->
+                  let+ _ = Build_system.build_file prog in
+                  (Path.to_string prog, args))
+        in
+        let argv = coqtop :: "-topfile" :: coq_file :: args @ extra_args in
```

Comment:
Here I'm using the original file path (as provided by the user) for the `-topfile` option, assuming `coqtop` is gonna be run in the current working directory. Is that the right thing to do?

Reply:
I don't think this works if the `.v` file is generated no?

Reply:
Yes indeed, but the user should be able to step through their local files (whether they have been written to disk or not), so I think that this is what we want. If the user wants to step through a generated file, their only choice with what I did is to edit the generated file directly with something like `dune coq top _build/default/.../generated.v`. Perhaps we want something special for generated files? But I think that from the perspective of the editor it is a bit weird to try to open a file that does not exist (since it only appears under `_build`).

Reply:
> whether they have been written to disk or not

How would that work? You've only compiled the dependencies for the saved `.v` source.

Reply:
There is indeed the issue of what to do if you add a new import to the file you are editing, and that new dependency has not been compiled before. I guess, in that case, you'd have no choice but to save and restart `coqtop`, or to build the dependency independently (e.g., with `dune build path/to/dep.vo`). That's not really any different to what currently happens when you work with a Coq makefile though. Any thoughts on that @ejgallego?

An alternative would be to build all the files of the theory (and its dependencies), but I don't things that's what you want in general, since running `coqc` can be quite slow.

Reply:
> An alternative would be to build all the files of the theory (and its dependencies), but I don't things that's what you want in general, since running coqc can be quite slow.

Yeah that definitely doesn't sound right.

Tbh, requiring users to save the file doesn't sound so bad to me. It's an easy rule to remember and keeps everything consistent.

I still don't understand the reasoning behind giving coqtop actual source paths. I understand it seems more natural as a first reaction, but are there concrete benefits? Apart from generated file, conditional compilation, etc, it has to potential to break the hygienic property of dune builds. In dune, we only allow access to source files under very controlled circumstances (promotion). Do we really want to break it over this?

Reply:
> I still don't understand the reasoning behind giving coqtop actual source paths.

As opposed to what? I guess you'd want paths under `_build`? We should, but for other reasons.

AFAICT (see test below and links) the `path` passed to `-topfile` need not exist and is not read, so @rgrinberg's concern about "access to source files" might not apply. It seems currently up to the editor/IDE to read the file, split it in statements, and feed them to `coqtop` on demand.

However, `-topfile path` lets `coqtop` compute the name of the module being processed, by looking up `path` in the loadpath computed from the `-Q` and `-R` options; this is on Github issues (https://github.com/ProofGeneral/PG/issues/407 https://github.com/coq/coq/pull/8991) and I reproduced this below. For instance, `-R _build/default/iris/ Bla -topfile _build/default/iris/foo.v` will cause `coqtop` to call the current module as `Bla.foo` (rather than the traditional `Top`). In this example, `-topfile _build/default/iris/foo.v` produces the wrong results — `-Q`, `-R` and `-topfile` must align. And since  `coqtop`'s `-Q`/`-R` arguments must point to folders in `_build` to find built `.vo` files, the argument of `-topfile` shall as well.

The listing below confirms that using the testcase from https://github.com/ProofGeneral/PG/issues/407#issuecomment-447289254!
In the listing below, we see that test fails under `rlwrap coqtop -R _build/default/iris/ Bla -topfile iris/foo.v`, but works under `rlwrap coqtop -R _build/default/iris/ Bla -topfile _build/default/iris/foo.v`.

```coq
$ rlwrap coqtop -R _build/default/iris/ Bla -topfile iris/foo.v
Welcome to Coq KillBill.local:/Users/pgiarrusso1/.opam/4.13.1+flambda-coq813-bump-tmp/.opam-switch/build/coq.8.13.2+flambda+bedrock,main (7c85aec48efff57cb43d8df9410c8d88e3a5f64d)

Coq < Definition hello := Prop.
hello is defined

Coq < Check Bla.foo.hello.
Toplevel input, characters 6-19:
> Check Bla.foo.hello.
>       ^^^^^^^^^^^^^
Error: The reference Bla.foo.hello was not found in the current environment.

Coq < About hello.
hello : Type

hello is not universe polymorphic
hello is transparent
Expands to: Constant foo.hello

Coq <

iris$ rlwrap coqtop -R _build/default/iris/ Bla -topfile _build/default/iris/foo.v
Welcome to Coq KillBill.local:/Users/pgiarrusso1/.opam/4.13.1+flambda-coq813-bump-tmp/.opam-switch/build/coq.8.13.2+flambda+bedrock,main (7c85aec48efff57cb43d8df9410c8d88e3a5f64d)

Coq < Definition hello := Prop.
hello is defined

Coq < Check Bla.foo.hello.
hello
     : Type

Coq <
```

Reply:
> AFAICT (see test below and links) the path passed to -topfile need not exist and is not read, so @rgrinberg's concern about "access to source files" might not apply. It seems currently up to the editor/IDE to read the file, split it in statements, and feed them to coqtop on demand.

It also means there's less of a reason to break dune's property. If the file isn't even read, why do we even care to pass the source path?

Reply:
OK, that makes sense, so file under `_build` it is then.

Reply:
(And as I understand @Blaisorblade's explanations, that's the only option.)

Reply:
Fixed in d76527569950b0b00f687b2befa7ca212aa48048.